### PR TITLE
ARROW-7622: [Format] Mark Tensor and SparseTensor fields required

### PR DESCRIPTION
**Pull request by pitrou**
Also make sparse COO indices row-major by default, both by consistency with tensors and because it packs related coordinates sequentially.

### CODE REVIEW

File: format/Tensor.fbs [hunk 1]
```diff
@@ -38,16 +38,16 @@ table TensorDim {
...
-  type: Type;
+  type: Type (required);
 
   /// The dimensions of the tensor, optionally named
-  shape: [TensorDim];
+  shape: [TensorDim] (required);
 
   /// Non-negative byte offsets to advance one value cell along each dimension
-  strides: [long];
+  strides: [long] (required);
```
Comment (wesm) [inline format/Tensor.fbs]: In the C++ API for tensors, `strides` is allowed to be not specified in which case the strides are computed according to C-order / row-major striding. Do we want to do the same here or require that the strides be set always in the protocol message?

**Review by wesm — COMMENTED** [review]
Minor comment but otherwise looks good. cc @mrkn for thoughts also

File: format/Tensor.fbs (hunk 1 above)
Comment (mrkn) [reply format/Tensor.fbs]: Both are fine to me. Allowing omission is useful #for applications which use C-major arrays without strides. In Arrow C++ case, we always have strides internally, so we won’t omit strides.

Comment (pitrou) [reply format/Tensor.fbs]: If we make it optional, should `indicesStrides` in `SparseTensorIndexCOO` also be optional?

Comment (mrkn) [reply format/Tensor.fbs]: Yes, we should also make it optional for the API consistency.

Comment (pitrou) [reply format/Tensor.fbs]: Ok. The problem is that the C++ code currently isn't ready for missing strides (it reads `indicesStrides` unconditionally).

Comment (wesm) [reply format/Tensor.fbs]: I see. I don't know the sparse C++ code well enough to fix myself at the moment

File: cpp/src/arrow/ipc/reader.cc [hunk 2]
```diff
@@ -871,17 +872,27 @@ Result<std::shared_ptr<SparseIndex>> ReadSparseCOOIndex(
...
   ARROW_ASSIGN_OR_RAISE(auto indices_data,
                         file->ReadAt(indices_buffer->offset(), indices_buffer->length()));
   std::vector<int64_t> indices_shape(
       {non_zero_length, static_cast<int64_t>(shape.size())});
   auto* indices_strides = sparse_index->indicesStrides();
-  std::vector<int64_t> strides;
-  // Assume indices_strides is a 2-length array.
-  strides.push_back(indices_strides->Get(0));
-  strides.push_back(indices_strides->Get(1));
+  std::vector<int64_t> strides(2);
```
Comment (pitrou) [inline cpp/src/arrow/ipc/reader.cc]: @mrkn Can you check if those changes look ok to you?

**Comment by pitrou** [comment]
@mrkn Is there a reason why Tensor strides default to row-major but SparseIndexCOO strides default to column-major?

**Comment by mrkn** [comment]
Actually, the default strides of SparseIndexCOO is column-major, but it isn't intentionally decided, just historically.
At first, SparseIndexCOO only supported column-major strides because scipy.sparse.coo_matrix always has a column-major index.  When I added the support of row-major, I didn't change the default strides.  That's all.

I think now is the best timing to change the default strides to row-major.

**Comment by pitrou** [comment]
Actually, I think it makes sense for SparseIndexCOO to be column-major. If it is row-major, then reading a column of coordinates yields non-sequential cache-unfriendly memory accesses. If it is column-major, the coordinates are sequential.

**Comment by pitrou** [comment]
(delete a message where I was making a mistake. Row-major is actually the preferred order which makes coordinates sequential in memory)

**Comment by pitrou** [comment]
I'll note that pydata/sparse uses row-major order for coordinates:
```python
>>> x = np.eye(4)                                                                                                                                                     
>>> x[2, 3] = 5                                                                                                                                                       
>>> s = sparse.COO.from_numpy(x)                                                                                                                                      
>>> s.coords                                                                                                                                                          
array([[0, 1, 2, 2, 3],
       [0, 1, 2, 3, 3]])
>>> s.coords.strides                                                                                                                                                  
(40, 8)
```

As for scipy.sparse, it only supports 2-dim sparse tensors and stores row and column coordinates in separate arrays... (which has the same memory access properties as column-major)

**Comment by pitrou** [comment]
Ok, I've switched to row-major by default for COO indices.

**Pull request closed without merging**